src/Qir/Runtime/lib/QIR/QubitManager.cpp

```diff
@@ -0,0 +1,468 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+// This code compiles, passes some unit tests, but isn't yet used in the runtime.
+
+#include "QubitManager.hpp"
+
+#include <cstring> // For memcpy
+
+namespace Microsoft
+{
+namespace Quantum
+{
+
+
+//
+// QubitListInSharedArray
+//
+
+QubitListInSharedArray::QubitListInSharedArray(QubitIdType startId, QubitIdType endId, QubitIdType* sharedQubitStatusArray)
+{
+    FailIf(startId > endId || startId < 0 || endId == std::numeric_limits<QubitIdType>::max(),
+        "Incorrect boundaries in the linked list initialization.");
+
+    for (QubitIdType i = startId; i < endId; i++) {
+        sharedQubitStatusArray[i] = i + 1; // Current element points to the next element.
+    }
+    sharedQubitStatusArray[endId] = NoneMarker; // Last element ends the chain.
+    firstElement = startId;
+    lastElement = endId;
+}
+
+bool QubitListInSharedArray::IsEmpty() const
+{
+    return firstElement == NoneMarker;
+}
+
+void QubitListInSharedArray::AddQubit(QubitIdType id, bool addToFront, QubitIdType* sharedQubitStatusArray)
+{
+    FailIf(id == NoneMarker, "Incorrect qubit id, cannot add it to the list.");
+
+    // If the list is empty, we initialize it with the new element.
+    if (IsEmpty())
+    {
+        firstElement = id;
+        lastElement = id;
+        sharedQubitStatusArray[id] = NoneMarker; // List with a single elemenet in the chain.
+        return;
+    }
+
+    if (addToFront)
+    {
+        sharedQubitStatusArray[id] = firstElement; // The new element will point to the former first element.
+        firstElement = id; // The new element is now the first in the chain.
+    } else
+    {
+        sharedQubitStatusArray[lastElement] = id; // The last element will point to the new element.
+        sharedQubitStatusArray[id] = NoneMarker; // The new element will end the chain.
+        lastElement = id; // The new element will be the last element in the chain.
+    }
+}
+
+// TODO: Set status to the taken qubit here. Then counting is reasonable here, but not possible?
+// TODO: Rename 'RemoveQubitFromFront'?
+int QubitListInSharedArray::TakeQubitFromFront(QubitIdType* sharedQubitStatusArray)
+{
+    // First element will be returned. It is 'NoneMarker' if the list is empty.
+    int result = firstElement;
```

Comment:
Forgot to replace `int` with `QubitIdType`? (2 occurrences)

Reply:
Thanks for catching! Updated.

Reply:
FYI: I see the update in .cpp but not in .hpp

Reply:
Thanks. Probably the last place that I overlooked.